File: busing-bridge/src/packet.rb
class Packet
  SIZE = 9
  NIL_PACKET = [0] * SIZE

  def initialize(bytes = NIL_PACKET)
    @bytes = bytes
  end

  def self.command(command_code)
    Packet.new(bytes = [255, 255, 0, 0, 0, 0, 0, 0, 0]).tap do |packet|
      packet.command = command_code
    end
  end

  def self.response(bytes)
    Packet.new(bytes)
  end

  def reception?
    @bytes[0] == 254 && @bytes[1] == 254
  end

  def command?
    @bytes[0] == 255 && @bytes[1] == 255
  end

  def response?
    @bytes[0] == 254 && @bytes[1] == 255
  end

  def none?
    @bytes.uniq == [0]
  end

  def type
    return "ACK" if reception? && command == 1
    return "RCV" if reception?
    return "CMD" if command?
    return "RSP" if response?
    return "NONE" if none?
    
    "UNKNOWN"
  end

  def valid?
    reception? || command? || response?
  end

  def address_to
    @bytes[3] * 256 + @bytes[4]
  end

  def address_to=(address)
    @bytes[3] = address / 256
    @bytes[4] = address % 256
  end

  def address_from
    return address_to if response?
    return "N/A" if command?

    @bytes[5] * 256 + @bytes[6]
  end

  def address_from=(address)
    @bytes[5] = address / 256
    @bytes[6] = address % 256
  end

  def command
    @bytes[2]
  end

  def command=(command)
    @bytes[2] = command
  end

  def data1
    @bytes[7]
  end

  def data1=(data)
    @bytes[7] = (data)
  end

  def data2
    @bytes[8]
  end

  def data2=(data)
    @bytes[8] = (data)
  end

  def as_command
    @bytes[0] = 255
    @bytes[1] = 255
    @bytes[5] = 0
    @bytes[6] = 0
    [
      @bytes[0],
      @bytes[1],
      @bytes[3],
      @bytes[4],
      @bytes[2],
      @bytes[7],
      @bytes[8]
    ]
  end

  def inspect
    {
      type: type,
      from: address_from,
      to: address_to,
      command: command,
      command_as_bit: command.chr.unpack('B*'),
      data1: data1,
      data1_as_bit: data1.chr.unpack('B*'),
      data2: data2,
      data2_as_bit: data2.chr.unpack('B*'),
      data_as_16_bit: data2 * 256 + data1,
      bytes: @bytes
    }
  end
end

File: busing-bridge/src/packet.py
"""Busing wire protocol: fixed-size 9-byte datagrams.

Byte layout of a received packet:

    0-1: frame type marker (254,254 = reception; 255,255 = command; 254,255 = response)
    2:   command code
    3-4: destination address (big endian)
    5-6: source address (big endian)
    7:   data1
    8:   data2

Packets are sent on the wire as 7 bytes (see ``as_command``).
"""


class Packet:
    SIZE = 9

    def __init__(self, data=None):
        self._bytes = list(data) if data is not None else [0] * self.SIZE

    @classmethod
    def for_command(cls, command_code):
        packet = cls([255, 255, 0, 0, 0, 0, 0, 0, 0])
        packet.command = command_code
        return packet

    def is_reception(self):
        return self._bytes[0] == 254 and self._bytes[1] == 254

    def is_command(self):
        return self._bytes[0] == 255 and self._bytes[1] == 255

    def is_response(self):
        return self._bytes[0] == 254 and self._bytes[1] == 255

    def is_none(self):
        return set(self._bytes) == {0}

    def is_valid(self):
        return self.is_reception() or self.is_command() or self.is_response()

    @property
    def type(self):
        if self.is_reception() and self.command == 1:
            return "ACK"
        if self.is_reception():
            return "RCV"
        if self.is_command():
            return "CMD"
        if self.is_response():
            return "RSP"
        if self.is_none():
            return "NONE"
        return "UNKNOWN"

    @property
    def command(self):
        return self._bytes[2]

    @command.setter
    def command(self, value):
        self._bytes[2] = value

    @property
    def address_to(self):
        return self._bytes[3] * 256 + self._bytes[4]

    @address_to.setter
    def address_to(self, address):
        self._bytes[3] = address // 256
        self._bytes[4] = address % 256

    @property
    def address_from(self):
        if self.is_response():
            return self.address_to
        if self.is_command():
            return "N/A"
        return self._bytes[5] * 256 + self._bytes[6]

    @address_from.setter
    def address_from(self, address):
        self._bytes[5] = address // 256
        self._bytes[6] = address % 256

    @property
    def data1(self):
        return self._bytes[7]

    @data1.setter
    def data1(self, value):
        self._bytes[7] = value

    @property
    def data2(self):
        return self._bytes[8]

    @data2.setter
    def data2(self, value):
        self._bytes[8] = value

    def as_command(self):
        """Serialize as the 7-byte on-wire command frame."""
        self._bytes[0] = 255
        self._bytes[1] = 255
        self._bytes[5] = 0
        self._bytes[6] = 0
        return bytes([
            self._bytes[0],
            self._bytes[1],
            self._bytes[3],
            self._bytes[4],
            self._bytes[2],
            self._bytes[7],
            self._bytes[8],
        ])

    def describe(self):
        return {
            "type": self.type,
            "from": self.address_from,
            "to": self.address_to,
            "command": self.command,
            "command_as_bit": format(self.command, "08b"),
            "data1": self.data1,
            "data1_as_bit": format(self.data1, "08b"),
            "data2": self.data2,
            "data2_as_bit": format(self.data2, "08b"),
            "data_as_16_bit": self.data2 * 256 + self.data1,
            "bytes": self._bytes,
        }

    def __repr__(self):
        return f"Packet({self.describe()})"
